### DISCUSSION TITLE
How to run docker command in this Airflow docker container?

### DISCUSSION THREAD

**Discussion by jeesim2** [apache/airflow#15378] [Q&A]
I have set up Airflow as described here. 
https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html

And there is a airflow task that have to execute docker command like 
```
BashOperator(
    task_id='my',
    bash_command="""
        docker run ..............
    """,
    dag=dag,
)
```

This means Docker package required [in the airflow docker image, but there is not](https://github.com/apache/airflow/blob/master/Dockerfile).

So, I tried to build my own `airflow image based docker installed image`. like,
```
FROM apache/airflow:2.0.1 as airflow-image
SHELL ["/bin/bash", "-o", "pipefail", "-e", "-u", "-x", "-c"]

RUN apt-get update
RUN apt-get install docker-ce docker-ce-cli containerd.io
```

But here was no luck. I couldn't build `airflow image based docker installed image`.
Actually I couldn't fully understood the [airflow Dockerfile](https://github.com/apache/airflow/blob/master/Dockerfile) and even `RUN apt-get update`, `RUN sudo apt-get update` failed.

Are you agree we can build `airflow image based docker installed image`? Or any suggestion to make docker cli available in airflow docker container? As you think I am a docker newbee.

--- 
## Update 
This Dockerfile successfully installed docker to the airflow docker base image
```
FROM apache/airflow:2.0.1 as airflow-image
SHELL ["/bin/bash", "-o", "pipefail", "-e", "-u", "-x", "-c"]

USER root
RUN apt-get update && apt-get --assume-yes install \
    apt-transport-https \
    ca-certificates \
    curl \
    gnupg \
    lsb-release
RUN curl -fsSL https://download.docker.com/linux/debian/gpg | sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
RUN echo $(lsb_release -cs)
RUN echo \
  "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/debian \
  $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
RUN apt-get update
RUN apt-get --assume-yes install docker-ce docker-ce-cli containerd.io systemd
RUN service docker start
```

But,  After I see this ENTRYPOINT and tested several things,
https://github.com/apache/airflow/blob/master/Dockerfile#L535
```
ENTRYPOINT ["/usr/bin/dumb-init", "--", "/entrypoint"]
```
I wonder whether I cant start docker daemon or not.
Any advice?

**Comment by potiuk** [comment]
This really depends on what you want to achieve.

Your installation process (the UPDATE) part is fine and it installs client. Docker consist of the `client` and `engine/daemon`. You are generally not supposed to 'just'  start docker daemon inside another docker container o, or you have to used -priviledged option when you start your container. 

You can connect to an external engine or the same docker engine your container runs on or even run a separate `dind` image which you connect to eventually, or you could build-in `dind` capabilities and run your container with -priviledged mode. Depending on your requirements and deployment (for example if you run in K8S or Docker-compose or others) you have different possibilities. 

And there are various pros/cons of those different approaches.

You can find a number of resources about running docker-in-docker - usually named dind (for example here: https://github.com/jpetazzo/dind and original blog post here https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/).
